tactic/basic.lean

```diff
@@ -452,29 +452,29 @@ do r ← result,
 /-- Succeeds only if the current goal is a proposition. -/
 meta def propositional_goal : tactic unit :=
 do goals ← get_goals,
-   let current_goal := goals.head,
-   current_goal_type ← infer_type current_goal,
-   p ← is_prop current_goal_type,
+   p ← is_proof goals.head,
    guard p
 
 variable {α : Type}
 
 private meta def iterate_aux (t : tactic α) : list α → tactic (list α)
 | L := (do r ← t, iterate_aux (r :: L)) <|> return L
 
+/-- Apply a tactic as many times as possible, collecting the results in a list. -/
 meta def iterate' (t : tactic α) : tactic (list α) :=
 list.reverse <$> iterate_aux t []
 
+/-- Like iterate', but fail if the tactic does not succeed at least once. -/
 meta def iterate1 (t : tactic α) : tactic (α × list α) :=
-do r ← t | fail "iterate1 failed: tactic did not succeed",
+do r ← decorate_ex "iterate1 failed: tactic did not succeed" t,
```

Comment:
The old `r <- t | fail ...` does nothing, because `|` is syntax for what to do on a failed pattern match and the pattern `r` can never fail. So it was just equivalent to `r <- t`. I thought it might be nicer to use `decorate_ex` here.